### ISSUE TITLE
Suspicious forcing of a lazy parameter and a discarded non-unit

### ISSUE THREAD

**Issue by alexknvl** [zio/zio#3173]
In `StreamEffect`:
```scala
  def fromIterator[A](iterator: => Iterator[A]): StreamEffect[Any, Nothing, A] =
    StreamEffect {
      // Discarded non-unit
      Managed.effectTotal(() => if (iterator.hasNext) iterator.next() else end)
      // Eagerly forced by-name parameter. StreamEffect.apply is eager in its param.
      val it = iterator
      Managed.effectTotal(() => if (it.hasNext) it.next() else end)
    }
```
@adamgfraser 

---
On a separate note, perhaps it's also a good time to address #2334

**Comment by adamgfraser** [comment]
I will look at this.

**Comment by adamgfraser** [comment]
Resolved by #3175.
